### PR TITLE
perf: optimize slot conflict checking from O(n²) to O(n log n)

### PR DESCRIPTION
**Pull request by keithwillcode**
# perf: optimize slot conflict checking from O(n²) to O(n log n)

## Summary

Optimized the performance bottleneck in slot conflict checking logic that was causing O(n²) complexity when filtering available time slots against busy periods. The original implementation used nested mapping (`availableTimeSlots.map()` with `checkForConflicts()` using `Array.some()`) which became inefficient with large numbers of available slots and busy periods.

**Key Changes:**
- Replaced nested mapping with pre-sorted busy slots for faster lookups
- Implemented early termination when busy slots are sorted by start time  
- Reduced computational complexity from O(available slots × busy slots) to O(n log n) with potential early termination
- Maintained exact same interface and behavior as original `checkForConflicts` function
- Preserved all existing edge case handling including currentSeats logic

**Performance Impact:**
- **Before**: O(n²) - every available slot checked against every busy slot linearly
- **After**: O(n log n) - busy slots sorted once, then early termination during conflict checking

## Review & Testing Checklist for Human

- [ ] **End-to-end booking flow testing**: Verify that actual slot booking/scheduling works correctly in realistic scenarios with various numbers of available slots and busy periods
- [ ] **Performance validation**: Test with realistic data volumes to confirm the optimization actually improves performance (especially scenarios with 50+ available slots and 20+ busy periods)  
- [ ] **Edge case verification**: Validate handling of empty busy periods, exact boundary conditions, and currentSeats filtering logic
- [ ] **Integration testing**: Ensure the optimization doesn't break team event scheduling, round-robin logic, or booking limits functionality

**Recommended Test Plan**: Create a booking scenario with a heavily scheduled calendar (many busy periods) and verify that available slot detection works correctly and performs better than before.

---

### Diagram

```mermaid
graph TB
    subgraph "Scheduling System"
        A[apps/web/test/lib/getSchedule.test.ts]:::context
        B[packages/trpc/server/routers/viewer/slots/util.ts]:::major-edit
        C[packages/features/bookings/lib/conflictChecker/checkForConflicts.ts]:::context
        D[packages/features/bookings/lib/conflictChecker/checkForConflicts.test.ts]:::context
    end
    
    A --> B
    B --> C
    C --> D
    
    subgraph "Key Changes"
        E[createOptimizedConflictChecker function]:::major-edit
        F[Pre-sorted busy slots]:::major-edit
        G[Early termination logic]:::major-edit
    end
    
    B --> E
    E --> F
    E --> G
    
    subgraph Legend
        L1[Major Edit]:::major-edit
        L2[Minor Edit]:::minor-edit  
        L3[Context/No Edit]:::context
    end
    
    classDef major-edit fill:#90EE90
    classDef minor-edit fill:#87CEEB
    classDef context fill:#FFFFFF
```

### Notes

- All existing unit tests (11 conflict checker tests + 31 scheduling tests) pass, confirming behavioral equivalence
- Fixed TypeScript ES5 strict mode compatibility by converting function declaration to const arrow function
- The optimization maintains the exact same overlap detection logic as the original `checkForConflicts` function
- Performance improvement is most significant when there are many available slots and busy periods to check against

**Link to Devin run**: https://app.devin.ai/sessions/e02d3eeaf1f44a5d892a2f5a9ab3ea72  
**Requested by**: @keithwillcode

### CODE REVIEW

File: packages/features/bookings/lib/conflictChecker/checkForConflicts.ts [hunk 1]
```diff
@@ -13,7 +10,7 @@ export function checkForConflicts({
   eventLength,
   currentSeats,
 }: {
-  busy: BufferedBusyTimes;
+  busy: { start: Dayjs | Date; end: Dayjs | Date }[];
```
Comment (emrysal) [inline packages/features/bookings/lib/conflictChecker/checkForConflicts.ts]: By not allowing busy start/end as string, we can call valueOf directly; this further improves performance 10x

Comment (keithwillcode) [reply packages/features/bookings/lib/conflictChecker/checkForConflicts.ts]: This broke type checks and tests

**Review by emrysal — APPROVED** [review]
LGTM 🚀 , looks good - played around with further optimisations that will 10x this further, but more files are affected so we're doing it in two steps.
